src/imageops/sample.rs

```diff
@@ -304,6 +304,71 @@ where
     out
 }
 
+/// Linearly bisample from an image using coordinates in [0,1].
+pub fn bilinear_sample<P: Pixel>(
+    img: &impl GenericImageView<Pixel = P>,
+    uv: [f32; 2],
+) -> Option<P> {
+    if !uv.iter().all(|c| (0.0..=1.0).contains(c)) {
```

Comment:
This method currently treats the center of the top left pixel as `(0, 0)`, while 3D graphics APIs generally specify that the center of that pixel is at `(0.5 / width, 0.5 / height)`.

Let's have the input range go from `(0.0, 0.0)` to `(width-1, height-1)` instead. That way we sidestep the question and avoid confusion. Conveniently it is already the range needed for `ui` and `vi`. 

In any case, there should be more documentation on the method explaining exactly how the coordinates map to pixels.

Reply:
do you mean the input range for uv? the point of having it be in the range 0-1 was so that it could be independent of the resolution of the image. Or do you mean the validation should be done on `ui` and `vi`?

This is intended to closely mirror the API of glsl's `texture` function. If we instead make the function coordinates in the  range `(0, w/h-1)`, then any user who operates in `0,1` space will have to multiply by the coordinate indices, which is a bit of a pain and can lead to subtle off by 1 errors (multiplying by w instead of w-1). My use case for these functions is I'm using it to sample textures from a mesh, and for that coordinates are entirely in 0-1 space

Reply:
Hm, matching GLSL's `texture` function does sound useful, especially given all the subtlety around pixel corners vs. centers. We could break the method into two pieces:
* `interpolate_bilinear` which used coordinates from (0, 0) to (width-1, height-1), and contained the actual interpolation logic, and
* `sample_bilinear` which uses (0, 0) to (1, 0) coordinates matching GLSL style, and is implemented by calling into the first method.

Reply:
Also, the proper conversion from [0,1] -> [0, width-1] is `(u * width - 0.5).max(0.0).min(width-1)`